### PR TITLE
Add tests for not abusing interfaces

### PR DESCRIPTION
**Pull request by keradus**
Inspired by #2396, #2405
Replacing them as well (at least in their main goal).

### CODE REVIEW

**Comment by keradus** [comment]
cc @SpacePossum

**Comment by SpacePossum** [comment]
nice idea :)
(needs some tweaking for 5.3 but I'm sure that will be alright)
Are you going to adopt the tests for public class properties as well?

**Comment by keradus** [comment]
added @SpacePossum

File: tests/ProjectCodeTest.php [hunk 1]
```diff
@@ -37,10 +37,16 @@ public function testThatSrcClassesNotAbuseInterfaces($className)
             ? new DocBlock($rc->getDocComment())
             : null;
 
+        // HHVM knows better which interfaces you implements
+        $interfaces = $rc->getInterfaces();
+        if (defined('HHVM_VERSION') && interface_exists('Stringish')) {
+            $interfaces = array_diff($interfaces, array('Stringish'));
+        }
```
Comment (keradus) [inline tests/ProjectCodeTest.php]: seriously...

File: tests/ProjectCodeTest.php [hunk 2]
```diff
@@ -26,6 +26,153 @@
...
      * @param string $className
      *
+     * @dataProvider provideSrcClasses
+     * @requires PHP 5.4
+     */
+    public function testThatSrcClassesNotAbuseInterfaces($className)
+    {
+        // HHVM knows better which interfaces you implements
+        if (defined('HHVM_VERSION') && interface_exists('Stringish')) {
+            $this->markTestSkipped('Skipped as HHVM violate inheritance tree with `Stringish` interface.');
```
Comment (keradus) [inline tests/ProjectCodeTest.php]: seriously...

File: tests/ProjectCodeTest.php [hunk 3]
```diff
@@ -26,6 +26,154 @@
...
      *
+     * @dataProvider provideSrcClasses
+     * @requires PHP 5.4
+     */
+    public function testThatSrcClassesNotAbuseInterfaces($className)
+    {
+        // HHVM knows better which interfaces you implements
+        // https://github.com/facebook/hhvm/issues/5890
+        if (defined('HHVM_VERSION') && interface_exists('Stringish')) {
+            $this->markTestSkipped('Skipped as HHVM violate inheritance tree with `Stringish` interface.');
```
Comment (keradus) [inline tests/ProjectCodeTest.php]: seriously...

**Pull request closed without merging**